File: dodevops-api/router/task/task.go
package task

import (
	"dodevops-api/api/task/controller"
	"dodevops-api/api/task/service"
	"dodevops-api/common"
	"dodevops-api/middleware"

	"github.com/gin-gonic/gin"
)

func RegisterTaskRoutes(router *gin.RouterGroup) {
	// 任务模板路由
	router.POST("/template/add", middleware.AuthMiddleware(), controller.CreateTemplate)
	router.GET("/template/list", middleware.AuthMiddleware(), controller.GetAllTemplates)
	router.PUT("/template/update", middleware.AuthMiddleware(), controller.UpdateTemplate)
	router.DELETE("/template/delete", middleware.AuthMiddleware(), controller.DeleteTemplate)
	router.GET("/template/info/:id", middleware.AuthMiddleware(), controller.GetTemplateByID)
	router.GET("/template/content/:id", middleware.AuthMiddleware(), controller.GetTemplateContent)
	router.GET("/template/query/name", middleware.AuthMiddleware(), controller.GetTemplatesByName)
	router.GET("/template/query/type", middleware.AuthMiddleware(), controller.GetTemplatesByType)

	// 任务管理路由
	router.POST("/task/add", middleware.AuthMiddleware(), controller.CreateTask)                        // 创建任务
	router.GET("/task/get", middleware.AuthMiddleware(), controller.GetTaskByID)                        // 获取任务信息
	router.PUT("/task/update", middleware.AuthMiddleware(), controller.UpdateTask)                      // 修改任务
	router.DELETE("/task/delete", middleware.AuthMiddleware(), controller.DeleteTask)                   // 删除任务
	router.GET("/task/list", middleware.AuthMiddleware(), controller.ListTasks)                         // 获取任务列表
	router.GET("/task/list-with-details", middleware.AuthMiddleware(), controller.ListTasksWithDetails) // 获取任务列表（包含关联信息）
	router.GET("/task/query/name", middleware.AuthMiddleware(), controller.GetTasksByName)              // 获取任务名称列表
	router.GET("/task/query/type", middleware.AuthMiddleware(), controller.GetTasksByType)              // 获取任务类型列表
	router.GET("/task/query/status", middleware.AuthMiddleware(), controller.GetTasksByStatus)          // 获取任务状态列表
	router.GET("/task/next-execution", middleware.AuthMiddleware(), controller.GetNextExecutionTime)    // 获取任务下次执行时间
	router.GET("/task/execution-info", middleware.AuthMiddleware(), controller.GetTaskExecutionInfo)    // 获取任务执行信息
	router.GET("/task/templates", middleware.AuthMiddleware(), controller.GetTaskTemplatesWithStatus)   // 获取任务模板列表

	// 任务作业路由
	router.POST("/taskjob/start", middleware.AuthMiddleware(), controller.TaskWork().StartJob)
	router.GET("/taskjob/log", middleware.AuthMiddleware(), controller.TaskWork().GetJobLog)
	router.POST("/taskjob/stop", middleware.AuthMiddleware(), controller.TaskWork().StopJob)
	router.GET("/taskjob/status", middleware.AuthMiddleware(), controller.TaskWork().GetJobStatus)

	// 任务监控路由
	taskMonitorCtrl := controller.NewTaskMonitorController()
	router.GET("/task/monitor/queue/metrics", middleware.AuthMiddleware(), taskMonitorCtrl.GetQueueMetrics)        // 获取队列指标
	router.GET("/task/monitor/scheduler/stats", middleware.AuthMiddleware(), taskMonitorCtrl.GetSchedulerStats)    // 获取调度器统计
	router.GET("/task/monitor/system/status", middleware.AuthMiddleware(), taskMonitorCtrl.GetSystemStatus)        // 获取系统状态
	router.GET("/task/monitor/queue/details", middleware.AuthMiddleware(), taskMonitorCtrl.GetQueueDetails)        // 获取队列详情
	router.POST("/task/monitor/queue/clear-failed", middleware.AuthMiddleware(), taskMonitorCtrl.ClearFailedQueue) // 清空失败队列
	router.POST("/task/monitor/queue/retry-failed", middleware.AuthMiddleware(), taskMonitorCtrl.RetryFailedTasks) // 重试失败任务

	// 定时任务管理路由
	router.POST("/task/monitor/scheduled/pause", middleware.AuthMiddleware(), taskMonitorCtrl.PauseScheduledTask)       // 暂停定时任务
	router.POST("/task/monitor/scheduled/resume", middleware.AuthMiddleware(), taskMonitorCtrl.ResumeScheduledTask)     // 恢复定时任务
	router.POST("/task/monitor/scheduled/reset", middleware.AuthMiddleware(), taskMonitorCtrl.ResetScheduledTaskStatus) // 重置定时任务状态
	router.GET("/task/monitor/task/status", middleware.AuthMiddleware(), taskMonitorCtrl.GetTaskStatus)                 // 获取任务状态详情

	// Ansible任务路由
	taskAnsibleCtrl := controller.NewTaskAnsibleController(service.NewTaskAnsibleService(common.GetDB()))
	router.GET("/task/ansiblelist", middleware.AuthMiddleware(), taskAnsibleCtrl.List)                   // 获取任务列表
	router.POST("/task/ansible", middleware.AuthMiddleware(), taskAnsibleCtrl.CreateTask)                // 创建Ansible任务
	router.POST("/task/k8s", middleware.AuthMiddleware(), taskAnsibleCtrl.CreateK8sTask)                 // 创建K8s任务
	router.GET("/task/ansible/:id", middleware.AuthMiddleware(), taskAnsibleCtrl.GetTask)                // 通过任务ID获取任务信息
	router.PUT("/task/ansible/:id", middleware.AuthMiddleware(), taskAnsibleCtrl.UpdateTask)             // 修改任务
	router.POST("/task/ansible/:id/start", middleware.AuthMiddleware(), taskAnsibleCtrl.StartTask)       // 启动任务(支持Ansible和K8s)
	router.DELETE("/task/ansible/:id", middleware.AuthMiddleware(), taskAnsibleCtrl.DeleteTask)          // 删除任务(级联删除子任务)
	router.GET("/task/ansible/:id/log/:work_id", middleware.AuthMiddleware(), taskAnsibleCtrl.GetJobLog) // 获取任务日志(SSE)
	router.GET("/task/ansible/query/name", middleware.AuthMiddleware(), taskAnsibleCtrl.GetTasksByName)  // 根据名称模糊查询任务
	router.GET("/task/ansible/query/type", middleware.AuthMiddleware(), taskAnsibleCtrl.GetTasksByType)  // 根据类型查询任务

	// Ansible配置管理路由
	configAnsibleCtrl := controller.NewConfigAnsibleController(service.NewConfigAnsibleService(common.GetDB()))
	router.POST("/config/ansible", middleware.AuthMiddleware(), configAnsibleCtrl.Create)
	router.PUT("/config/ansible/:id", middleware.AuthMiddleware(), configAnsibleCtrl.Update)
	router.DELETE("/config/ansible/:id", middleware.AuthMiddleware(), configAnsibleCtrl.Delete)
	router.GET("/config/ansible/:id", middleware.AuthMiddleware(), configAnsibleCtrl.Get)
	router.GET("/config/ansible", middleware.AuthMiddleware(), configAnsibleCtrl.List)

}

// RegisterWebSocketRoutes 注册WebSocket路由（不需要中间件认证）
func RegisterWebSocketRoutes(router *gin.RouterGroup) {
	// WebSocket任务日志路由 (内部处理认证)
	wsCtrl := controller.NewWebSocketController(service.NewTaskAnsibleService(common.GetDB()))
	router.GET("/ws/task/ansible/:id/log/:work_id", wsCtrl.GetJobLogWS) // WebSocket日志接口
}

File: web/src/api/task.js
import request from '@/utils/request'
import storage from "@/utils/storage"

export function CreateTemplate(data) {
  return request({
    url: 'template/add',
    method: 'post',
    data,
    headers: {
      'Content-Type': 'application/json'
    }
  })
}

export function GetAllTemplates(params) {
  return request({
    url: 'template/list',
    method: 'get',
    params,
    headers: {
      'Accept': 'application/json'
    }
  })
}

export function UpdateTemplate(data) {
  return request({
    url: `template/update?id=${data.id}`,
    method: 'put',
    data: {
      name: data.name,
      type: data.type,
      content: data.content,
      remark: data.remark
    }
  })
}

export function DeleteTemplate(params) {
  return request({
    url: 'template/delete',
    method: 'delete',
    params
  })
}
// 获取模板
export function GetTemplateByID(params) {
  return request({
    url: 'template/info/' + params.id,
    method: 'get'
  })
}

export function GetTemplateContent(params) {
  return request({
    url: 'template/content/' + params.id,
    method: 'get',
    headers: {
      'Accept': 'text/plain'
    }
  })
}

export function GetTemplatesByName(params) {
  console.log('查询模板名称参数:', params)
  return request({
    url: `template/query/name?name=${encodeURIComponent(params.name || '')}`,
    method: 'get',
    params: {
      pageNum: params.pageNum,
      pageSize: params.pageSize
    },
    headers: {
      'Accept': 'application/json',
      'Authorization': localStorage.getItem('token') || ''
    }
  })
}

export function GetTemplatesByType(params) {
  console.log('查询模板类型参数:', params)
  return request({
    url: `template/query/type?type=${params.type}`,
    method: 'get',
    params: {
      pageNum: params.pageNum,
      pageSize: params.pageSize
    },
    headers: {
      'Accept': 'application/json',
      'Authorization': localStorage.getItem('token') || ''
    }
  })
}

// 任务管理API
export function CreateTask(data) {
  return request({
    url: 'task/add',
    method: 'post',
    data,
    headers: {
      'Content-Type': 'application/json'
    }
  })
}

export function GetTaskByID(params) {
  return request({
    url: 'task/get',
    method: 'get',
    params
  })
}

export function UpdateTask(data) {
  return request({
    url: 'task/update',
    method: 'put',
    data
  })
}

export function DeleteTask(data) {
  return request({
    url: 'task/delete',
    method: 'delete',
    data,
    headers: {
      'Content-Type': 'application/json'
    }
  })
}

export function ListTasks(params) {
  return request({
    url: 'task/list',
    method: 'get',
    params
  })
}

export function GetTasksByName(params) {
  return request({
    url: 'task/query/name',
    method: 'get',
    params
  })
}

export function GetTasksByType(params) {
  return request({
    url: 'task/query/type',
    method: 'get',
    params
  })
}

export function GetTasksByStatus(params) {
  return request({
    url: 'task/query/status',
    method: 'get',
    params
  })
}

export function GetNextExecutionTime(params) {
  return request({
    url: 'task/next-execution',
    method: 'get',
    params
  })
}

// 新增获取任务模板接口
export function GetTaskTemplates(params) {
  console.log('获取任务模板参数:', params)
  return request({
    url: 'task/templates',
    method: 'get',
    params: {
      id: params.id
    },
    headers: {
      'Accept': 'application/json',
      'Authorization': localStorage.getItem('token') || ''
    }
  }).catch(error => {
    console.error('获取任务模板失败:', error)
    throw error
  })
}

export function GetTaskJobLog(params) {
  return request({
    url: 'taskjob/log',
    method: 'get',
    params: {
      taskId: params.taskId,
      templateId: params.templateId
    },
    headers: {
      'Accept': 'application/json'
    }
  })
}

export function StartJob(data) {
  return request({
    url: `taskjob/start?taskId=${data.taskId}`,
    method: 'post',
    headers: {
      'accept': 'application/json'
    }
  })
}

export function StopJob(params) {
  return request({
    url: 'taskjob/stop',
    method: 'post',
    params: {
      taskId: params.taskId,
      templateId: params.templateId
    },
    headers: {
      'Content-Type': 'application/x-www-form-urlencoded',
      'Accept': 'application/json'
    }
  })
}

// 获取Ansible配置列表
export function GetAnsibleConfigList(params) {
  return request({
    url: '/config/ansible',
    method: 'get',
    params
  })
}

// 创建Ansible配置
export function CreateAnsibleConfig(data) {
  return request({
    url: '/config/ansible',
    method: 'post',
    data,
    headers: {
      'Content-Type': 'application/json'
    }
  })
}

// 更新Ansible配置
/**
 * 
 * body参数示例：
 {
  "content": "string",
  "name": "string",
  "remark": "string",
  "type": 0
}
 */
export function UpdateAnsibleConfig(data) {
  return request({
    url: `/config/ansible/${data.id}`,
    method: 'put',
    data,
    headers: {
      'Content-Type': 'application/json'
    }
  })
}

// 获取Ansible配置详情
export function GetAnsibleConfigById(id) {
  return request({
    url: `/config/ansible/${id}`,
    method: 'get'
  })
}

// 删除Ansible配置
export function DeleteAnsibleConfig(id) {
  return request({
    url: `/config/ansible/${id}`,
    method: 'delete'
  })
}

// Ansible任务管理API
export function GetAnsibleTaskList(params) {
  return request({
    url: 'task/ansiblelist',
    method: 'get',
    params
  })
}

export function CreateAnsibleTask(data) {
  return request({
    url: 'task/ansible',
    method: 'post',
    data,
    headers: {
      'Content-Type': 'multipart/form-data'
    }
  })
}

export function UpdateAnsibleTask(data) {
  return request({
    url: `task/ansible/${data.id}`,
    method: 'put',
    data,
    headers: {
      'Content-Type': 'application/json'
    }
  })
}

export function GetAnsibleTaskById(id) {
  return request({
    url: `task/ansible/${id}`,
    method: 'get'
  })
}

export function StartAnsibleTask(id) {
  return request({
    url: `task/ansible/${id}/start`,
    method: 'post'
  })
}

/**
 * 建立Ansible任务实时日志的SSE连接
 * @param {string|number} id - 任务ID
 * @param {string|number} workId - 执行ID
 * @returns {EventSource} 返回EventSource实例，调用者需自行添加 onmessage 监听和 close 处理
 */
export function GetAnsibleTaskLog(id, workId, retryCount = 0) {
  const maxRetries = 1 // 减少重试次数，避免过长等待
  // 根据后端处理时间调整：首次30秒，重试3分钟
  const timeout = retryCount === 0 ? 30000 : 180000

  return request({
    url: `/task/ansible/${id}/log/${workId}`,
    method: 'get',
    timeout: timeout,
    params: {
      // 添加时间戳避免缓存
      t: Date.now(),
      // 请求实时日志，即使任务在运行中
      realtime: true,
      // 包含缓冲区日志
      includeBuffer: true
    }
  }).then(response => {
    // 记录实际的API调用和响应
    console.log('🔍 日志API调用详情:', {
      taskId: id,
      workId: workId,
      requestUrl: `/api/v1/task/ansible/${id}/log/${workId}`,
      timestamp: new Date().toISOString(),
      responseSize: response.data ? JSON.stringify(response.data).length : 0,
      responsePreview: response.data ? JSON.stringify(response.data).substring(0, 100) + '...' : 'null',
      retryCount: retryCount,
      timeout: timeout / 1000 + 's'
    })
    return response
  }).catch(error => {
    console.error('❌ 获取历史日志失败:', error)

    // 如果是超时错误且未达到最大重试次数，进行重试


    // 抛出带有更多上下文信息的错误
    const timeoutInfo = retryCount === 0 ? '30秒' : '3分钟'
    const enhancedError = {
      ...error,
      contextMessage: `后端处理超过${timeoutInfo}，可能正在执行长时间操作`
    }
    throw enhancedError
  })
}

// 获取Ansible任务日志历史记录列表
export function GetAnsibleTaskHistory(params) {
  return request({
    url: `/api/v1/task/ansible/${params.id}/history`,
    method: 'get',
    params: {
      page: params.page,
      limit: params.pageSize,
    }
  })
}

// 根据任务ID、WORKID和HistoryID获取历史任务日志
export function GetAnsibleTaskLogByHistory(params) {
  return request({
    url: `/api/v1/task/ansible/history/detail/task/${params.id}/work/${params.workId}/history/${params.historyId}/log`,
    method: 'get',
    params: {
      t: Date.now(),
      // 获取历史日志时不需要实时参数
      realtime: false,
      includeBuffer: false
    }
  })
}

// 获取任务的历史执行详情，包含每个主机的执行日志
export function GetAnsibleHistoryDetail(params) {
  return request({
    url: `/api/v1/task/ansible/${params.id}/history/${params.historyId}`,
    method: 'get',
    headers: {
      'Accept': 'application/json'
    }
  })
}

// 尝试直接获取日志文件内容的备选方法
export function GetAnsibleTaskLogDirect(id, workId) {
  console.log('🔧 尝试直接日志文件访问方法')
  return request({
    url: `/task/ansible/${id}/log/${workId}/direct`,
    method: 'get',
    timeout: 10000, // 10秒超时
    params: {
      t: Date.now(),
      // 强制读取最新日志
      tail: true,
      // 获取最后1000行
      lines: 1000,
      // 绕过缓存
      nocache: true
    }
  }).catch(error => {
    console.warn('直接日志访问失败，使用标准方法:', error.message)
    throw error
  })
}

// 获取有效token的工具函数
function getValidToken() {
  // 尝试多种存储位置获取token
  const storageKeys = ['token', 'access_token', 'jwt_token', 'authToken']
  let token = null

  // 优先从localStorage获取
  for (const key of storageKeys) {
    token = storage.getItem(key)
    if (token && token !== 'null' && token !== 'undefined') {
      break
    }
  }

  // 如果localStorage没有，尝试sessionStorage
  if (!token || token === 'null' || token === 'undefined') {
    for (const key of storageKeys) {
      token = sessionStorage.getItem(key)
      if (token && token !== 'null' && token !== 'undefined') {
        break
      }
    }
  }

  // 如果token是JSON对象，尝试解析
  if (token && typeof token === 'string' && token.startsWith('{')) {
    try {
      const tokenObj = JSON.parse(token)
      token = tokenObj.token || tokenObj.access_token || tokenObj.value || tokenObj.jwt
    } catch (e) {
      console.warn('无法解析token JSON:', e)
    }
  }

  // 确保token不是字符串'null'
  if (token === 'null' || token === 'undefined' || !token) {
    token = null
  }

  console.log('Token获取结果:', {
    hasToken: !!token,
    tokenPreview: token ? `${token.substring(0, 10)}...` : 'null',
    tokenLength: token ? token.length : 0
  })

  return token
}

// SSE实时日志流接口（回调式）
export function GetAnsibleTaskLogStream(id, workId, handlers = {}) {
  const { onOpen, onMessage, onError, onClose } = handlers
  const token = storage.getItem("token") || {}
  const baseURL = (process.env.VUE_APP_API_BASE_URL || '').replace(/\/$/, '')
  const url = `${baseURL}/api/v1/task/ansible/${id}/log/${workId}?t=${Date.now()}&realtime=true&includeBuffer=true`

  const controller = new AbortController()
  const decoder = new TextDecoder()

  fetch(url, {
    method: 'GET',
    headers: {
      'Authorization': token ? `Bearer ${token}` : '',
      'Accept': 'text/event-stream'
    },
    signal: controller.signal
  })
    .then(async response => {
      if (!response.ok) {
        throw new Error(`HTTP ${response.status}: ${response.statusText}`)
      }

      onOpen && onOpen()

      const reader = response.body.getReader()
      let buffer = ''

      while (true) {
        const { done, value } = await reader.read()
        if (done) break

        buffer += decoder.decode(value, { stream: true })
        const events = buffer.split('\n\n')
        buffer = events.pop() || ''

        events.forEach(eventBlock => {
          eventBlock.split('\n').forEach(line => {
            if (line.startsWith('data:')) {
              const payload = line.slice(5).trim()
              if (payload) {
                onMessage && onMessage(payload)
              }
            }
          })
        })
      }

      onClose && onClose()
    })
    .catch(error => {
      if (error.name !== 'AbortError') {
        console.error('❌ 日志请求失败:', error)
        onError && onError(error)
      }
      onClose && onClose()
    })

  return {
    close() {
      controller.abort()
    }
  }
}

// 获取Ansible任务详情和Works列表
export function GetAnsibleTaskDetail(id) {
  return request({
    url: `task/ansible/${id}`,
    method: 'get',
    headers: {
      'Accept': 'application/json'
    }
  })
}

// 启动Ansible任务
export function StartAnsibleTaskFlow(id) {
  return request({
    url: `task/ansible/${id}/start`,
    method: 'post',
    headers: {
      'Accept': 'application/json'
    }
  })
}

export function DeleteAnsibleTask(id) {
  return request({
    url: `task/ansible/${id}`,
    method: 'delete',
    headers: {
      'Accept': 'application/json'
    }
  })
}

// 根据名称模糊查询Ansible任务
export function GetAnsibleTasksByName(params) {
  console.log('查询Ansible任务名称参数:', params)
  return request({
    url: `task/ansible/query/name?name=${encodeURIComponent(params.name || '')}`,
    method: 'get',
    params: {
      page: params.page,
      pageSize: params.pageSize
    },
    headers: {
      'Accept': 'application/json',
      'Authorization': localStorage.getItem('token') || ''
    }
  }).catch(error => {
    console.error('按名称查询Ansible任务失败:', error)
    throw error
  })
}

// 根据类型查询Ansible任务
export function GetAnsibleTasksByType(params) {
  console.log('查询Ansible任务类型参数:', params)
  return request({
    url: `task/ansible/query/type?type=${params.type}`,
    method: 'get',
    params: {
      page: params.page,
      pageSize: params.pageSize
    },
    headers: {
      'Accept': 'application/json',
      'Authorization': localStorage.getItem('token') || ''
    }
  }).catch(error => {
    console.error('按类型查询Ansible任务失败:', error)
    throw error
  })
}

// 暂停定时任务
export function PauseScheduledTask(taskId) {
  return request({
    url: 'task/monitor/scheduled/pause',
    method: 'post',
    params: {
      task_id: taskId
    },
    headers: {
      'Accept': 'application/json'
    }
  })
}

// 恢复定时任务
export function ResumeScheduledTask(taskId) {
  return request({
    url: 'task/monitor/scheduled/resume',
    method: 'post',
    params: {
      task_id: taskId
    },
    headers: {
      'Accept': 'application/json'
    }
  })
}
